### PR TITLE
proceed to state 13, family code wrong

### PR DESCRIPTION
**Pull request by qiguaiderenmen**
Error messages:

DS2480B status = 16ffe09b
DS2480B state = 13
Unknown device family code: 0x16 (DS2505 should be 0x0B)

### CODE REVIEW

**Comment by qiguaiderenmen** [comment]
A new update: adding a read state after sending 0x33 (Read ROM Command) solves the issue. Now the error message becomes:

DS2480B status = bffe0db
DS2480B state = 13
Failed to read instrument memory

**Comment by qiguaiderenmen** [comment]
Update: change several states and now proceed to write data to .txt file. Unfortunately, we only receive 0xFF for all the data.

**Comment by qiguaiderenmen** [comment]
Update: code refactored with fewer states, 'processor' based functional blocks and more fluent logic flow. More comments involved. Still within all 0xFF data issue, perhaps due to no proper 1-wire configuration involved.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: I am not sure this will work as intended.  Does it just initialize ds_program[7:8] with the current contents of mem_addr (presumably 0)?

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: Why was this state changed?  It is only used for the direct 1-wire interface, which it seems you have broken.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: Same issue.  You are breaking the direct 1-wire interface

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: Your state names are confusing.  You are using 1_WIRE for states that are used with the DS2480B chip.  Note that the alternative (current working implementation) is the direct 1-wire interface, so this could lead to some confusion.  Perhaps it would be better to use DS2480B instead of 1_WIRE.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: This line not needed.  `unexpected_idx` will be set in the next state.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: I'm not sure this will update `ds_program`.  Instead of updating `mem_data`, you could directly update `ds_program[7:8]`

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [reply FPGA1394_QLA/Verilog/DS2505.v]: More importantly, you should set `progCnt` rather than assuming it will be 5, because once the DS2480B is configured, future calls will jump directly to `DS_RESET_1_WIRE` and in that case `progCnt` will probably not be 5.

**Review by pkazanzides — CHANGES_REQUESTED** [review]
This is looking better.  One major issue is that I believe you have broken the direct 1-wire interface.  Making sure that still works is a requirement before merging into the devel branch.

I think the best next step is to completely separate the states related to the direct 1-wire interface and the DS2480B interface. Note also my comment about naming of your states. Using 1_WIRE for states related to the DS2480B (and not the direct 1-wire interface) is extremely confusing. 

Then, make sure the direct 1-wire interface still works.  At the end, we could consider consolidating states between the DS2480B and direct 1-wire interface, as Shi Xin had done to some extent, though I'm not sure it would give much benefit.

Another minor point is that you can also add `DS_RESET_1_WIRE`, `DS_RESP_1_WIRE`, and `DS_SET_DATA_MODE` to the program.  The only "special cases" are the setting of `DS480B_ok` and `ds_reset`, which could be handled differently.  The simplest (though perhaps not best) solution is to set them based on the value of `progCnt`.

I am also thinking that eventually the `DS_RUN_READ_PREP_FUNC_BLK` and `DS_CHECK_BYTE_READ_PREP` could be combined with the "config" states, but that can wait.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (qiguaiderenmen) [reply FPGA1394_QLA/Verilog/DS2505.v]: Yes professor, I've changed the statements from 'initial' to 'assign'.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (qiguaiderenmen) [reply FPGA1394_QLA/Verilog/DS2505.v]: These parts are just ignored for now, will merge them once the ds2480 functionality is ready.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (qiguaiderenmen) [reply FPGA1394_QLA/Verilog/DS2505.v]: Yeah, same to previous one.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (qiguaiderenmen) [reply FPGA1394_QLA/Verilog/DS2505.v]: Thanks for pointing out! The name has been changed to DS_WAIT_2MS_FOR_2480_RESET, as well as other 2ms delay states.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (qiguaiderenmen) [reply FPGA1394_QLA/Verilog/DS2505.v]: Yeah that's a huge bug! `progCnt` is now assigned to 5 and the program looks much more solid.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (qiguaiderenmen) [reply FPGA1394_QLA/Verilog/DS2505.v]: No professor, this part has no relationship to `ds_program`, from my perspective. It is continuously sending 0xFF and storing the received bytes to mem_data (in the most recent update I reduce the number to 8 bytes, but still get all ff). If you mean mem_addr, I think it's updated in the very beginning of DS_IDLE state.

Comment (pkazanzides) [reply FPGA1394_QLA/Verilog/DS2505.v]: Right. I had confused mem_data with mem_addr.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: I assume this is temporary. You should not be initializing to 1.

**Comment by qiguaiderenmen** [comment]
In commit ee4fd44, Jie Ying's 5 original files are included (to distinguish from current serial code they're modified to ds_<name>).
In commit cc6b4b9, some modifications are made, but we can't read anything from the chip. Maybe due to wrong settings in FPGA1394Eth-QLA.v, not pretty sure.

**Comment by qiguaiderenmen** [comment]
Works! I've replaced the fragile Read state with my own receive implementation. It works just fine and I'll do some code refine & adjustment with direct wire interface.

**Comment by qiguaiderenmen** [comment]
Commit `f382161` is the working version with old serial write implementation, commit `6895d22` is in new style but doesn't work.

**Comment by qiguaiderenmen** [comment]
Serial trial for both old / new QLA & 1-wire trial for new QLA have been tested and validated. Everything works just fine.

File: FPGA1394_QLA/Verilog/uart_recv_2480.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/uart_recv_2480.v]: Why not replace this case statement by `rxdata[rx_cnt-1] <= uart_rxd_d1`?

File: FPGA1394_QLA/Verilog/uart_recv_2480.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/uart_recv_2480.v]: Not needed.

File: FPGA1394_QLA/Verilog/uart_recv_2480.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/uart_recv_2480.v]: Replace tabs by spaces in this file.

File: FPGA1394_QLA/CMakeLists.txt
Comment (pkazanzides) [inline FPGA1394_QLA/CMakeLists.txt]: I see there is also a uart_send_2480.v.  Should that be added to CMakeLists.txt?

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v:465]: This should not be hard-coded.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v:527]: Why is there uart code here too?

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v:128]: Could add an extra 2 bits and eliminate `DS_CHECK_PROGRAMMER`.  The simplest is to add 5 bits for the next state, but there really are only 3 possible next states, so 2 bits would be enough.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: Change this to `wire`. Then, `assign tx_data[9] = 1; assign tx_data[8:1] = out_byte; assign tx_data[0] = 0;`

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: Could change to `out_byte <= 8'h00;` (assuming change to `tx_data` described above)

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: As above, change from `tx_data` to `out_byte`

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v:452]: Could do something like `next_state <= next_state_array[ds_program[progCnt][17:16]];`, assuming `next_state_array` is properly set up to specify `DS_PROGRAMMER`, `DS_READ_MEM_REQUEST` or `DS_IDLE`.  Might also needs to set `num_bytes <= 8'd0` here.  Then, can eliminate `DS_CHECK_PROGRAMMER`.

File: FPGA1394_QLA/Verilog/DS2505.v
Comment (pkazanzides) [inline FPGA1394_QLA/Verilog/DS2505.v]: Again, use `out_byte`.

**Review by pkazanzides — CHANGES_REQUESTED** [review]
Please clean up the serial port code. I see some serial port code in DS2505.v and then there are files uart_recv_2480.v and uart_send_2480.v.  Please delete the code that is not being used.  Also, change tabs to spaces.

File: FPGA1394_QLA/Verilog/uart_recv_2480.v
Comment (pkazanzides) [reply FPGA1394_QLA/Verilog/uart_recv_2480.v]: Even better, based on what I think you are doing, would be a right shift, where the new data comes in on the left:  `rxdata <= { uart_rxd_d1, rxdata[6:0] };` (that is just one implementation).  You'll have to tweak the code a little to properly handle `rx_cnt == 0`, which I think currently does nothing.

**Review by pkazanzides — APPROVED** [review]
A few things left to do, such as reading the family code, but this is good enough to merge now.

**Pull request merged**
